webui/react/src/components/FilterForm/components/FilterGroup.tsx

```diff
@@ -0,0 +1,202 @@
+import { HolderOutlined, PlusOutlined } from '@ant-design/icons';
+import { Dropdown, DropDownProps } from 'antd';
+import type { MenuProps } from 'antd';
+import { useMemo, useRef } from 'react';
+import { useDrag, useDrop } from 'react-dnd';
+
+import ConjunctionContainer from 'components/FilterForm/components/ConjunctionContainer';
+import FilterField from 'components/FilterForm/components/FilterField';
+import { FilterFormStore, ITEM_LIMIT } from 'components/FilterForm/components/FilterFormStore';
+import { Conjunction, FormField, FormGroup, FormKind } from 'components/FilterForm/components/type';
+import Button from 'components/kit/Button';
+import Icon from 'components/kit/Icon';
+import { V1ProjectColumn } from 'services/api-ts-sdk';
+
+import css from './FilterGroup.module.scss';
+
+interface Props {
+  conjunction: Conjunction;
+  index: number; // start from 0
+  group: FormGroup;
+  parentId: string;
+  level: number; // start from 0
+  formStore: FilterFormStore;
+  columns: V1ProjectColumn[];
+}
+
+const FilterGroup = ({
+  index,
+  conjunction,
+  group,
+  level,
+  formStore,
+  parentId,
+  columns,
+}: Props): JSX.Element => {
+  const scrollBottomRef = useRef<HTMLDivElement>(null);
+  const [, drag, preview] = useDrag<{ form: FormGroup; index: number }, unknown, unknown>(() => ({
+    item: { form: group, index },
+    type: FormKind.Group,
+  }));
+
+  const [{ isOverCurrent, canDrop }, drop] = useDrop<
+    { form: FormGroup | FormField; index: number },
+    unknown,
+    { isOverCurrent: boolean; canDrop: boolean }
+  >({
+    accept: [FormKind.Group, FormKind.Field],
+    canDrop(item, monitor) {
+      const isOverCurrent = monitor.isOver({ shallow: true });
+      if (isOverCurrent) {
+        if (item.form.kind === FormKind.Group) {
+          return (
+            // cant self dnd
+            group.id !== item.form.id &&
+            // cant dnd in self childrens group
+            item.form.children.filter((c) => c.id === group.id).length === 0 &&
+            // cant dnd with deeper than 2 level group
+            level < 2 &&
+            // cant dnd if sum of source children of group type (0 if none, 1 if children exist)
+            // and target item's level is over 2
+            // 2 is the max depth
+            (item.form.children.filter((c) => c.kind === FormKind.Group).length === 0 ? 0 : 1) +
+              level <
+              2
+          );
+        }
+        return true;
+      }
+      return false;
```

Comment:
very tiny nit: maybe instead use a boolean statement to represent this so we understand why we're returning `true` or `false` near the end?
```suggestion
      return isOverCurrent && (item.form.kind !== FormKind.Group || (
            // cant self dnd
            group.id !== item.form.id &&
            // cant dnd in self childrens group
            item.form.children.filter((c) => c.id === group.id).length === 0 &&
            // cant dnd with deeper than 2 level group
            level < 2 &&
            // cant dnd if sum of source children of group type (0 if none, 1 if children exist)
            // and target item's level is over 2
            // 2 is the max depth
            (item.form.children.filter((c) => c.kind === FormKind.Group).length === 0 ? 0 : 1) +
              level <
              2
      ))
```

Reply:
it looks concise but it doesnt narrow the `form. kind` type due to `item.form.kind !== FormKind.Group`

Reply:
you are right. i probably replaced a wrong place